src/nvim/drawscreen.c

```diff
@@ -2326,9 +2320,12 @@ static void win_update(win_T *wp)
       idx++;
       lnum += foldinfo.fi_lines + 1;
     } else {
-      if (wp->w_p_rnu && wp->w_last_cursor_lnum_rnu != wp->w_cursor.lnum) {
-        // 'relativenumber' set and cursor moved vertically: The
-        // text doesn't need to be drawn, but the number column does.
+      // If:
+      // - 'number' is set and below inserted/deleted lines, or
+      // - 'relativenumber' is set and cursor moved vertically,
+      // the text doesn't need to be redrawn, but the number column does.
+      if ((wp->w_p_nu && mod_top != 0 && lnum >= mod_bot && buf->b_mod_xlines != 0)
```

Comment:
Should `b_mod_xlines` be reset somewhere after redrawing? Currently we always update the number column upon cursor movement after lines have been inserted/deleted, until an edit is made that doesn't insert/delete lines.

Reply:
Hmm, a buffer can be shown in multiple windows, so maybe that should be done in `update_screen()`.